### PR TITLE
feat: add the ability to use custom interactions

### LINKED ISSUE #757: Custom Interactions
### Summary

a `get_context()` alternative for slash command Interaction objects

### What is the feature request for?

Application commands

### The Problem

Currently, custom functions cannot be applied to `Interaction` objects like they can to the `Context` object used in normal commands

### The Ideal Solution

A `get_application()` method gets added to the `Client` object.

### The Current Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Skelmis**
## Summary

Adds the ability to provide a custom interaction class which the library will use.

Closes #757 

Note the return type of `get_interaction` varies from it's `get_context` counterpart as this was the only way I could find that would appease `task pyright`. If you are aware of an alternative where the return type can simply be `InterT` please let me know.

## Checklist

- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
  - [x] I have run `task pyright` and fixed the relevant issues.
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

## Test code

```python
import os

from nextcord import Interaction, Client

TESTING_GUILD_ID = 500525882226769931


class CustomInter(Interaction):
    @property
    def custom_prop(self) -> bool:
        return True


class CustomClient(Client):
    def get_interaction(self, data, *, cls=Interaction):
        return super().get_interaction(data, cls=CustomInter)


client = CustomClient()


@client.event
async def on_ready():
    print("...")


@client.slash_command(guild_ids=[TESTING_GUILD_ID])
async def test(interaction: CustomInter):
    await interaction.send(f"{isinstance(interaction, CustomInter)=}, {interaction.custom_prop}")


client.run(os.environ["TOKEN"])
```

### CODE REVIEW

File: nextcord/client.py [hunk 1]
```diff
@@ -2652,3 +2655,40 @@ def parse_mentions(self, text: str) -> List[User]:
...
+            class be provided, it should be a subclass of :py:class:`Interaction`.
+
+        Returns
+        -------
+        Interaction
+            An instance :py:class:`Interaction` or the provided subclass.
+
+        Notes
+        -----
+        This is synchronous due to how slash commands are implemented.
```
Comment (ooliver1) [inline nextcord/client.py:2692]: So it shows in a blue box like other notes
```suggestion
        .. note::

            This is synchronous due to how slash commands are implemented.
```

**Review by ooliver1 — CHANGES_REQUESTED** [review]
It might make sense to add an example on how to use this. In the docs or here in `./examples`

**Comment by DenverCoder1** [comment]
> If you are aware of an alternative where the return type can simply be `InterT` please let me know.

You can do it the dpy way, by adding `# type: ignore` on the return line.

**Comment by Skelmis** [comment]
> > If you are aware of an alternative where the return type can simply be `InterT` please let me know.
> 
> You can do it the dpy way, by adding `# type: ignore` on the return line.

Ideally without just ignoring things.

**Comment by Skelmis** [comment]
I couldn't think of a better example for a custom interaction, but it should serve it's purpose none the less

**Review by ooliver1 — APPROVED** [review]

**Review by teaishealthy — APPROVED** [review]

File: examples/custom_interaction.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+class MyInteraction(nextcord.Interaction):
+    async def send_embed(self, *, title: str, description: str):
+        """Easily send a basic embed."""
+        embed: nextcord.Embed = nextcord.Embed(
+            title=title,
+            description=description,
+        )
+        embed.set_author(name=self.user.display_name, icon_url=self.user.display_avatar)
+        await self.send(embed=embed)
```
Comment (DenverCoder1) [inline examples/custom_interaction.py:12]: I'd probably make this return the partialinteractionmessage/webhookmessage if I were doing this, but it's just an example, so it's fine either way

**Review by DenverCoder1 — APPROVED** [review]
Looks good, tested

**Review by ArjunSharda — APPROVED** [review]

**Pull request merged**